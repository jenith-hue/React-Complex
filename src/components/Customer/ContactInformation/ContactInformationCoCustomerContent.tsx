import {
  RACButton,
  RACTable,
  RACTableCell,
  RACTableRow,
  RACTooltip,
  Typography,
} from '@rentacenter/racstrap';
import React, { useState } from 'react';
import { PhoneInstructionsModal } from './PhoneInstructionsModal';
import { withContactStyles, WrappedComponentProps } from './withContactStyles';
import { abbrev, formatPhoneNumber } from '../../../utils/utils';
import { ApiStateWrapper } from '../../../common/ApiStateWrapper/ApiStateWrapper';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import clsx from 'clsx';

const coCustomerContentTestId = 'coCustomerContentTestId';

const ContactInformationCoCustomerContent_ = ({
  classes,
  onLogClicked,
  onError,
}: WrappedComponentProps) => {
  const {
    coCustomerDetails,
    loadingCoCustomer,
    hasApiErrorCoCustomer,
    isLogActivityAllowed,
  } = useCustomerDetails();

  const [showPhoneInstructionModal, setShowPhoneInstructionModal] =
    useState<boolean>(false);

  const [phoneInstructionsPhoneId, setPhoneInstructionsPhoneId] =
    useState<string>('');
  const [phoneInstructionsNotes, setPhoneInstructionsNotes] =
    useState<string>('');

  const popUpPhoneInstructions = (
    mykey: string,
    myvalue: string | null
  ): void => {
    setPhoneInstructionsPhoneId(mykey);
    setPhoneInstructionsNotes(myvalue == null ? '' : myvalue);
    setShowPhoneInstructionModal(true);
  };

  const closePhoneInstructionModal = () => {
    setShowPhoneInstructionModal(false);
  };

  const updatePhoneNotes = (phoneId: string, notes: string) => {
    coCustomerDetails?.phones?.forEach((one) => {
      if (phoneId == one.phoneId) {
        one.note = notes;
      }
    });
  };

  const renderTableHead = () => (
    <>
      <RACTableCell>Name</RACTableCell>
      <RACTableCell>Phone Number</RACTableCell>
      <RACTableCell>Type</RACTableCell>
      <RACTableCell>Best Time to Call</RACTableCell>
      <RACTableCell>Phone Instructions</RACTableCell>
      <RACTableCell># Contacts Today</RACTableCell>
      <RACTableCell>Last Call Result</RACTableCell>
      <RACTableCell align="center">Action</RACTableCell>
    </>
  );

  // eslint-disable-next-line sonarjs/cognitive-complexity
  const renderTableContent = () => (
    <>
      {coCustomerDetails &&
        coCustomerDetails.phones?.map((phone) => (
          <RACTableRow
            backgroundColor="white"
            key={phone.phoneId || phone.personalReferenceId}
          >
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {phone.phoneId
                  ? `${coCustomerDetails.firstName || ''} ${
                      coCustomerDetails.lastName || ''
                    }`
                  : `${phone.firstName || ''} ${phone.lastName || ''}`}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {formatPhoneNumber(
                  phone.phoneNumber?.split('.')[0],
                  !isLogActivityAllowed
                )}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {phone.phoneTypeDesc || phone.relationshipTypeDesc}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography variant="body2" className={classes.tableContentColor}>
                {phone.callTimeTypeDesc || phone.bestTimeToCallDesc}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              {showPhoneInstructionModal && (
                <PhoneInstructionsModal
                  phoneId={phoneInstructionsPhoneId}
                  note={phoneInstructionsNotes}
                  onChange={updatePhoneNotes}
                  onClose={closePhoneInstructionModal}
                  onError={onError}
                />
              )}
              {phone.phoneId && (
                <RACButton
                  variant="text"
                  size="small"
                  color="primary"
                  onClick={() =>
                    popUpPhoneInstructions(phone?.phoneId, phone?.note)
                  }
                >
                  {phone.note ? (
                    <Typography noWrap variant="body2" color="primary">
                      <RACTooltip title={phone?.note}>
                        <span>{abbrev(phone?.note)}</span>
                      </RACTooltip>
                    </Typography>
                  ) : (
                    <Typography
                      variant="body2"
                      color="primary"
                      className={classes.underline}
                    >
                      Add
                    </Typography>
                  )}
                </RACButton>
              )}
            </RACTableCell>
            <RACTableCell>
              <Typography
                align="center"
                variant="body2"
                className={classes.tableContentColor}
              >
                {phone.communicationsToday}
              </Typography>
            </RACTableCell>
            <RACTableCell>
              <Typography noWrap variant="body2" color="primary">
                <RACTooltip title={phone.lastCallResultDescription || ''}>
                  <span>{abbrev(phone.lastCallResultDescription || '')}</span>
                </RACTooltip>
              </Typography>
            </RACTableCell>
            <RACTableCell className={classes.center}>
              <RACButton
                variant="outlined"
                color="primary"
                className={classes.logButton}
                onClick={() =>
                  onLogClicked(
                    String(coCustomerDetails.customerId),
                    phone.phoneNumber
                  )
                }
                disabled={!phone.phoneNumber || !isLogActivityAllowed}
              >
                Log
              </RACButton>
            </RACTableCell>
          </RACTableRow>
        ))}
    </>
  );
  return (
    <div
      className={clsx(
        classes.listWrapper,
        (loadingCoCustomer || hasApiErrorCoCustomer) && classes.tabContentLoader
      )}
      data-testid={coCustomerContentTestId}
    >
      <ApiStateWrapper
        loading={loadingCoCustomer}
        hasApiError={hasApiErrorCoCustomer}
        response={coCustomerDetails}
        successContent={
          <RACTable
            renderTableHead={renderTableHead}
            renderTableContent={renderTableContent}
          />
        }
      />
    </div>
  );
};

export const ContactInformationCoCustomerContent = (props: any) =>
  withContactStyles(props)(ContactInformationCoCustomerContent_);
