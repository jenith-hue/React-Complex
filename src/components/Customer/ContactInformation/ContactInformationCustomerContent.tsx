import {
  RACButton,
  RACTable,
  RACTableCell,
  RACTableRow,
  RACTooltip,
  Typography,
} from '@rentacenter/racstrap';

import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { CustomerLocationState, Phone } from '../../../types/types';
import { EditContactInformationModal } from './EditContactInformationModal';
import { withContactStyles, WrappedComponentProps } from './withContactStyles';
import { abbrev, formatPhoneNumber } from '../../../utils/utils';
import clsx from 'clsx';
import { useTextConversationActions } from '../../../context/TextConversationProvider/TextConversationProvider';

const ContactInformationCustomerContent_ = ({
  classes,
  onLogClicked,
}: WrappedComponentProps) => {
  const {
    customerDetails: { phones, firstName, lastName },
    isLogActivityAllowed,
  } = useCustomerDetails();
  const location = useLocation<CustomerLocationState>();

  const [
    showEditCustomerInformationModal,
    setShowEditCustomerInformationModal,
  ] = useState<boolean>(false);

  const [tempPhone, setTempPhone] = useState<Phone>();
  const popUpPhoneInstructions = (phone: Phone): void => {
    setTempPhone(phone);
    setShowEditCustomerInformationModal(true);
  };

  const closePhoneInstructionModal = () => {
    setShowEditCustomerInformationModal(false);
  };
  const { updatePhoneOptionsfromEdit } = useTextConversationActions();
  const updatePhoneNotes = (
    phoneId: string,
    notes: string,
    bestTimeToCall: string,
    phoneNumber: string,
    phoneType: string,
    ext: string
  ) => {
    phones?.map((one) => {
      if (phoneId == one.phoneId) {
        one.note = notes;
        one.callTimeTypeDesc = bestTimeToCall;
        one.phoneNumber = phoneNumber;
        one.extension = ext;
        one.phoneTypeDesc = phoneType;
      }
    });
    if (phoneType === 'Cell Phone') {
      updatePhoneOptionsfromEdit();
    }
  };
  const renderTableHead = () => (
    <>
      <RACTableCell>Name</RACTableCell>
      <RACTableCell>Phone Number</RACTableCell>
      <RACTableCell>Type</RACTableCell>
      <RACTableCell>Best Time to Call</RACTableCell>
      <RACTableCell>Phone Instructions</RACTableCell>
      <RACTableCell># Contacts Today</RACTableCell>
      <RACTableCell>Last Call Result</RACTableCell>
      <RACTableCell align="center">Action</RACTableCell>
    </>
  );

  const customerId =
    location?.state?.customer?.customerId || location?.pathname?.split('/')[3];

  const renderTableContent = () => (
    <>
      {phones?.map((phone) => (
        <RACTableRow backgroundColor="white" key={phone.phoneId}>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {`${firstName} ${lastName}`}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {formatPhoneNumber(
                phone.phoneNumber?.split('.')[0],
                !isLogActivityAllowed
              )}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {phone.phoneTypeDesc}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {phone.callTimeTypeDesc}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <RACTooltip placement="right-start" title={phone?.note || ''}>
              <Typography
                noWrap
                variant="body2"
                color="primary"
                className={clsx(phone.note)}
              >
                <span>{phone.note ? abbrev(phone.note) : ''}</span>
              </Typography>
            </RACTooltip>
            <Typography noWrap variant="body2" color="primary"></Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography
              align="center"
              variant="body2"
              className={classes.tableContentColor}
            >
              {phone.communicationsToday}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <RACTooltip
              placement="right-start"
              title={phone.lastCallResultDescription || ''}
            >
              <Typography noWrap variant="body2" color="primary">
                <span>{abbrev(phone.lastCallResultDescription || '')}</span>
              </Typography>
            </RACTooltip>
          </RACTableCell>
          <RACTableCell className={classes.center}>
            <RACButton
              variant="outlined"
              color="primary"
              className={classes.logButton}
              onClick={() =>
                onLogClicked(customerId, phone.phoneNumber?.split('.')[0])
              }
              disabled={!phone?.phoneNumber || !isLogActivityAllowed}
            >
              Log
            </RACButton>
            <RACButton
              variant="contained"
              color="primary"
              className={classes.editButton}
              onClick={() => popUpPhoneInstructions(phone)}
            >
              Edit
            </RACButton>
          </RACTableCell>
        </RACTableRow>
      ))}
    </>
  );
  return (
    <>
      <EditContactInformationModal
        open={showEditCustomerInformationModal}
        phoneId={tempPhone?.phoneId || ''}
        note={tempPhone?.note || ''}
        ext={tempPhone?.extension || ''}
        bestTimeToCall={tempPhone?.callTimeType || ''}
        phoneType={tempPhone?.phoneType || ''}
        phoneNumber={tempPhone?.phoneNumber || ''}
        onChange={updatePhoneNotes}
        onClose={closePhoneInstructionModal}
      />
      <RACTable
        renderTableHead={renderTableHead}
        renderTableContent={renderTableContent}
      />
    </>
  );
};

export const ContactInformationCustomerContent = (props: any) =>
  withContactStyles(props)(ContactInformationCustomerContent_);
