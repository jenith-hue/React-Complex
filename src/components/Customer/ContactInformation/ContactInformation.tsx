import {
  makeStyles,
  RACCard,
  RACTabs,
  Typography,
  RACCOLOR,
} from '@rentacenter/racstrap';
import React, { useState } from 'react';
import { ContactInformationCustomerContent } from './ContactInformationCustomerContent';
import { ContactInformationReferencesContent } from './ContactInformationReferencesContent';
import { ContactInformationEmployerContent } from './ContactInformationEmployerContent';
import { ContactInformationCoCustomerContent } from './ContactInformationCoCustomerContent';
import { LogActivityModal } from './LogActivityModal';
import { useWorkedHistoryActions } from '../../../context/WorkedHistory/WorkedHistoryProvider';
import {
  ACCT_ACTIVITY_REF_CODES,
  LOG_ACTIVITY_ERROR_MESSAGE,
} from '../../../constants/constants';
import { ApiErrorModal } from '../../../common/ApiError/ApiError';
import { useCustomerPaymentSummary } from '../../../context/CustomerPaymentSummary/CustomerPaymentSummaryProvider';
import {
  useCustomerDetails,
  useCustomerDetailsActions,
} from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { formatDateString } from '../../../utils/utils';

const AVAILABLE_TABS = ['Customer', 'Co-Customer', 'References', 'Employer'];
const CO_CUSTOMER_TAB_INDEX = 1;
const REFERNCES_TAB_INDEX = 2;
const EMPLOYER_TAB_INDEX = 3;
/*
Reason for having both whiteSpace: 'nowrap' and <Typography noWrap ....>:
The one from css (th and td see below) will make the string to not break into new line.
The one from Typography will add the 3 dots (...) when string cannot be completely displayed.
*/
const useStyles = makeStyles((theme: any) => ({
  contactInformationRoot: {
    marginTop: '2rem',
    marginBottom: '1rem',
  },
  row: {
    display: 'flex',
    width: '100%',
    justifyContent: 'space-between',
    marginTop: theme.typography.pxToRem(8),
  },
  // TODO Align with designers regarding the card height
  // Should have a fixed height? Should be the same height as for text messages?
  // Or height depends on the number of rows?
  cardContentContainer: {
    marginLeft: '1.25rem',
    marginRight: '1.25rem',
    overflowX: 'auto',
    '& th': {
      whiteSpace: 'nowrap',
    },
    '& td': {
      maxWidth: theme.typography.pxToRem(160),
      whiteSpace: 'break-spaces',
    },
  },
  tabPanel: {
    height: '26rem',
    maxHeight: '26rem',
    overflow: 'auto',
  },
  dueContainer: {
    display: 'flex',
    gap: theme.typography.pxToRem(8),
  },
  dueInfo: {
    paddingRight: theme.typography.pxToRem(8),
  },
  dueKey: {
    color: `${RACCOLOR.MAXIMUM_RED} !important`,
  },
  dueValue: {
    color: `${RACCOLOR.CARNELIAN} !important`,
  },
}));

export const ContactInformation = () => {
  const classes = useStyles();

  const [openLogActivityModal, setOpenLogActivityModal] = useState(false);
  // can be the customer, co-customer, employer or reference
  const [personId, setPersonId] = useState('');
  const [callType, setCallType] = useState('');
  const [phoneNo, setPhoneNo] = useState('');

  const [isApiErrorOpen, setIsApiErrorOpen] = useState(false);
  const [apiErrorMessage, setApiErrorMessage] = useState('');

  const { onLogWorkedHistory } = useWorkedHistoryActions();

  const {
    customerDetails: { coCustomerId, personalReferences, employerReferences },
  } = useCustomerDetails();

  const { customerPaymentSummary: customerPaymentSummaryResponse } =
    useCustomerPaymentSummary();

  const { fetchCommunicationDetailsForCustomer } = useCustomerDetailsActions();

  const disabledTabsToStore = [];
  if (!coCustomerId) {
    disabledTabsToStore.push(CO_CUSTOMER_TAB_INDEX);
  }
  if (!personalReferences?.length) {
    disabledTabsToStore.push(REFERNCES_TAB_INDEX);
  }
  if (!employerReferences?.length) {
    disabledTabsToStore.push(EMPLOYER_TAB_INDEX);
  }

  const onLogWorkedHistoryWithSuccess = () => {
    setOpenLogActivityModal(false);
    setApiErrorMessage('');
  };

  const onLogWorkedHistoryWithError = () => {
    setOpenLogActivityModal(false);
    setIsApiErrorOpen(true);
    setApiErrorMessage(LOG_ACTIVITY_ERROR_MESSAGE);
  };

  const daysPastDue = customerPaymentSummaryResponse?.pastDueSummary
    ?.daysPastDue
    ? customerPaymentSummaryResponse?.pastDueSummary?.daysPastDue
    : '';

  const handleError = (message: string) => {
    setApiErrorMessage(message);
    setIsApiErrorOpen(true);
  };

  return (
    <div className={classes.contactInformationRoot}>
      {isApiErrorOpen && (
        <ApiErrorModal
          open={isApiErrorOpen}
          message={apiErrorMessage}
          onClose={() => {
            setApiErrorMessage('');
            setIsApiErrorOpen(false);
          }}
        />
      )}
      <div className={classes.row}>
        <Typography noWrap variant="h4">
          Contact Information
        </Typography>
        <div className={classes.dueContainer}>
          <div className={classes.dueInfo}>
            <Typography
              display="inline"
              variant="body1"
              className={classes.dueKey}
            >
              Due Date:{' '}
            </Typography>
            <Typography
              display="inline"
              variant="h6"
              className={classes.dueValue}
            >
              {formatDateString(
                customerPaymentSummaryResponse?.pastDueSummary?.pastDueDate
              )}
            </Typography>
          </div>
          <div className={classes.dueInfo}>
            <Typography
              display="inline"
              variant="body1"
              className={classes.dueKey}
            >
              Days Past Due:{' '}
            </Typography>
            <Typography
              display="inline"
              variant="h6"
              className={classes.dueValue}
            >
              {daysPastDue}
            </Typography>
          </div>
          <div className={classes.dueInfo}>
            <Typography
              display="inline"
              variant="body1"
              className={classes.dueKey}
            >
              Amount Due:{' '}
            </Typography>
            <Typography
              display="inline"
              variant="h5"
              className={classes.dueValue}
            >
              ${customerPaymentSummaryResponse?.pastDueSummary?.amountDue || ''}
            </Typography>
          </div>
        </div>
      </div>
      <RACCard>
        <div className={classes.cardContentContainer}>
          {openLogActivityModal && (
            <LogActivityModal
              open={openLogActivityModal}
              onSave={(callResult: string, note: string) => {
                onLogWorkedHistory(
                  callResult,
                  note,
                  daysPastDue?.toString(),
                  personId,
                  callType,
                  phoneNo,
                  onLogWorkedHistoryWithSuccess,
                  onLogWorkedHistoryWithError
                ).then(() => {
                  fetchCommunicationDetailsForCustomer();
                });
              }}
              onClose={() => setOpenLogActivityModal(false)}
            />
          )}
          <RACTabs
            classes={{
              tabPanel: classes.tabPanel,
            }}
            defaultValue={0}
            tabs={AVAILABLE_TABS}
            contentForTabs={[
              <ContactInformationCustomerContent
                key="customer"
                onLogClicked={(personId: string, phone: string) => {
                  setPersonId(personId);
                  setCallType(ACCT_ACTIVITY_REF_CODES.CALL_CUSTOMER);
                  setOpenLogActivityModal(true);
                  setPhoneNo(phone);
                }}
                onError={handleError}
              />,
              <ContactInformationCoCustomerContent
                key="co-customer"
                onLogClicked={(personId: string, phone: string) => {
                  setPersonId(personId);
                  setCallType(ACCT_ACTIVITY_REF_CODES.CALL_ALTERNATIVE);
                  setOpenLogActivityModal(true);
                  setPhoneNo(phone);
                }}
                onError={handleError}
              />,
              <ContactInformationReferencesContent
                key="references"
                onLogClicked={(personId: string, phone: string) => {
                  setPersonId(personId);
                  setCallType(ACCT_ACTIVITY_REF_CODES.CALL_REFERENCE);
                  setOpenLogActivityModal(true);
                  setPhoneNo(phone);
                }}
              />,
              <ContactInformationEmployerContent
                key="employer"
                onLogClicked={(personId: string, phone: string) => {
                  setPersonId(personId);
                  setCallType(ACCT_ACTIVITY_REF_CODES.CALL_EMPLOYER);
                  setOpenLogActivityModal(true);
                  setPhoneNo(phone);
                }}
              />,
            ]}
            // drop ts ignore after deploying racstrap
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            //@ts-ignore
            disabledTabs={disabledTabsToStore}
          />
        </div>
      </RACCard>
    </div>
  );
};
