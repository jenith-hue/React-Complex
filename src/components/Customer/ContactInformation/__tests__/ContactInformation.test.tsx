/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { ContactInformation } from '../ContactInformation';
import * as WorkedHistoryProvider from '../../../../context/WorkedHistory/WorkedHistoryProvider';

interface PropWithChildren {
  children: JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  ...jest.requireActual('@rentacenter/racstrap'),
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACCard: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACCard" {...rest}>
      {children}
    </div>
  ),
  RACTabs: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTabs" {...rest}>
      {children}
    </div>
  ),
  RACSelect: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACSelect" {...rest}>
      {children}
    </div>
  ),
  RACModal: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACModal" {...rest}>
      {children}
    </div>
  ),
  makeStyles: () => () => ({
    contactInformationRoot: 'contactInformationRoot',
    row: 'row',
    cardContentContainer: 'cardContentContainer',
    tabPanel: 'tabPanel',
  }),
  RACCOLOR: {
    ALICE_BLUE: 'ALICE_BLUE',
  },
}));

jest.mock('../ContactInformationReferencesContent', () => ({
  ContactInformationReferencesContent: ({
    children,
    ...rest
  }: PropWithChildren) => (
    <div data-testid="ReferencesContent" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../LogActivityModal', () => ({
  LogActivityModal: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="LogActivityModal" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock(
  '../../../../context/CustomerDetails/CustomerDetailsProvider',
  () => ({
    useCustomerDetails: () => ({
      customerDetails: {},
    }),
    useCustomerDetailsActions: () => ({
      fetchCommunicationDetailsForCustomer: jest.fn(),
      fetchCommunicationDetailsForCoCustomer: jest.fn(),
    }),
  })
);

/*
Observations:
- you cannot mock ContactInformationCustomerContent, ContactInformationCoCustomerContent, etc
(maybe because those are passed as props to RACTabs).

- if you add values to customerDetails mock, the tests will hang
*/
describe('ContactInformation', () => {
  it('renders correctly', async () => {
    const tree = renderer.create(<ContactInformation />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('renders with log activity modal', async () => {
    jest
      .spyOn(WorkedHistoryProvider, 'useWorkedHistoryActions')
      .mockImplementation(() => ({
        fetchWorkedHistory: jest.fn(),
        setReloadWorkedHistory: jest.fn(),
        onLogWorkedHistory: (
          callResult: string,
          note: string,
          daysPastDue: string,
          personId: string,
          callType: string,
          phoneNo: string,
          onSuccess: () => void,
          onError: () => void
        ): Promise<void> => {
          onError();
          return new Promise(jest.fn());
        },
      }));
    const tree = renderer.create(<ContactInformation />);

    renderer.act(() => {
      tree.root
        .findByProps({ 'data-testid': 'RACTabs' })
        .props.contentForTabs[0].props.onLogClicked('id');
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('renders with api error modal', async () => {
    jest
      .spyOn(WorkedHistoryProvider, 'useWorkedHistoryActions')
      .mockImplementation(() => ({
        fetchWorkedHistory: jest.fn(),
        setReloadWorkedHistory: jest.fn(),
        onLogWorkedHistory: (
          callResult: string,
          note: string,
          daysPastDue: string,
          personId: string,
          callType: string,
          phoneNo: string,
          onSuccess: () => void,
          onError: () => void
        ): Promise<void> => {
          onError();
          return new Promise(jest.fn());
        },
      }));
    const tree = renderer.create(<ContactInformation />);

    renderer.act(() => {
      tree.root
        .findByProps({ 'data-testid': 'RACTabs' })
        .props.contentForTabs[0].props.onLogClicked('id');
    });

    renderer.act(() => {
      tree.root
        .findByProps({ 'data-testid': 'LogActivityModal' })
        .props.onSave();
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
