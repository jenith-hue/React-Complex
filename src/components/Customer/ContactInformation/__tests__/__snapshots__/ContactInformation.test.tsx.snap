// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInformation renders correctly 1`] = `
<div
  className="contactInformationRoot"
>
  <div
    className="row"
  >
    <div
      data-testid="Typography"
      noWrap={true}
      variant="h4"
    >
      Contact Information
    </div>
    <div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Due Date:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h6"
        >
          
        </div>
      </div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Days Past Due:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h6"
        >
          
        </div>
      </div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Amount Due:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h5"
        >
          $
          
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="RACCard"
  >
    <div
      className="cardContentContainer"
    >
      <div
        classes={
          Object {
            "tabPanel": "tabPanel",
          }
        }
        contentForTabs={
          Array [
            <ContactInformationCustomerContent
              onError={[Function]}
              onLogClicked={[Function]}
            />,
            <ContactInformationCoCustomerContent
              onError={[Function]}
              onLogClicked={[Function]}
            />,
            <ContactInformationReferencesContent
              onLogClicked={[Function]}
            />,
            <ContactInformationEmployerContent
              onLogClicked={[Function]}
            />,
          ]
        }
        data-testid="RACTabs"
        defaultValue={0}
        disabledTabs={
          Array [
            1,
            2,
            3,
          ]
        }
        tabs={
          Array [
            "Customer",
            "Co-Customer",
            "References",
            "Employer",
          ]
        }
      />
    </div>
  </div>
</div>
`;

exports[`ContactInformation renders with api error modal 1`] = `
<div
  className="contactInformationRoot"
>
  <div
    buttons={
      <React.Fragment>
        <Button
          color="primary"
          onClick={[Function]}
          variant="contained"
        >
          OK
        </Button>
      </React.Fragment>
    }
    classes={
      Object {
        "dialog": undefined,
        "dialogActions": undefined,
        "dialogContent": undefined,
      }
    }
    content={
      <ApiErrorModalContent
        message="Unable to log activity. Please try again later."
      />
    }
    data-testid="RACModal"
    isOpen={true}
    maxWidth="sm"
    onClose={[Function]}
    title=""
    titleVariant="h4"
  />
  <div
    className="row"
  >
    <div
      data-testid="Typography"
      noWrap={true}
      variant="h4"
    >
      Contact Information
    </div>
    <div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Due Date:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h6"
        >
          
        </div>
      </div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Days Past Due:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h6"
        >
          
        </div>
      </div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Amount Due:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h5"
        >
          $
          
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="RACCard"
  >
    <div
      className="cardContentContainer"
    >
      <div
        classes={
          Object {
            "tabPanel": "tabPanel",
          }
        }
        contentForTabs={
          Array [
            <ContactInformationCustomerContent
              onError={[Function]}
              onLogClicked={[Function]}
            />,
            <ContactInformationCoCustomerContent
              onError={[Function]}
              onLogClicked={[Function]}
            />,
            <ContactInformationReferencesContent
              onLogClicked={[Function]}
            />,
            <ContactInformationEmployerContent
              onLogClicked={[Function]}
            />,
          ]
        }
        data-testid="RACTabs"
        defaultValue={0}
        disabledTabs={
          Array [
            1,
            2,
            3,
          ]
        }
        tabs={
          Array [
            "Customer",
            "Co-Customer",
            "References",
            "Employer",
          ]
        }
      />
    </div>
  </div>
</div>
`;

exports[`ContactInformation renders with log activity modal 1`] = `
<div
  className="contactInformationRoot"
>
  <div
    className="row"
  >
    <div
      data-testid="Typography"
      noWrap={true}
      variant="h4"
    >
      Contact Information
    </div>
    <div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Due Date:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h6"
        >
          
        </div>
      </div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Days Past Due:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h6"
        >
          
        </div>
      </div>
      <div>
        <div
          data-testid="Typography"
          display="inline"
          variant="body1"
        >
          Amount Due:
           
        </div>
        <div
          data-testid="Typography"
          display="inline"
          variant="h5"
        >
          $
          
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="RACCard"
  >
    <div
      className="cardContentContainer"
    >
      <div
        data-testid="LogActivityModal"
        onClose={[Function]}
        onSave={[Function]}
        open={true}
      />
      <div
        classes={
          Object {
            "tabPanel": "tabPanel",
          }
        }
        contentForTabs={
          Array [
            <ContactInformationCustomerContent
              onError={[Function]}
              onLogClicked={[Function]}
            />,
            <ContactInformationCoCustomerContent
              onError={[Function]}
              onLogClicked={[Function]}
            />,
            <ContactInformationReferencesContent
              onLogClicked={[Function]}
            />,
            <ContactInformationEmployerContent
              onLogClicked={[Function]}
            />,
          ]
        }
        data-testid="RACTabs"
        defaultValue={0}
        disabledTabs={
          Array [
            1,
            2,
            3,
          ]
        }
        tabs={
          Array [
            "Customer",
            "Co-Customer",
            "References",
            "Employer",
          ]
        }
      />
    </div>
  </div>
</div>
`;
